### PR TITLE
fix(lsp): only disable inlay hints / diagnostics if no other clients are connected

### PR DESCRIPTION
**Pull request by catlee**
In https://github.com/neovim/neovim/pull/24469#issuecomment-1651992691, @stasjok observed that in the current implementation, if one LSP client disconnects from a buffer, then inlay_hints and diagnostics will be disabled.

In this PR I've tried to fix that by examining the other LSP clients attached to the current buffer in the `LspDetach` handler. NB the handler is called *before* the client detaches, so we expect the number of attached clients to be 1 if this is the last client in the process of disconnecting.

I also used a similar method in each feature's refresh logic to prevent making calls if no attached clients support the method. This can happen if e.g. you enable inlay hints on a buffer that doesn't have any clients connected that support it.

### CODE REVIEW

File: runtime/lua/vim/lsp/diagnostic.lua
Comment (mfussenegger) [inline runtime/lua/vim/lsp/diagnostic.lua:14]: There was just another PR merged that adds constants in lsp.protocol.Methods

File: runtime/lua/vim/lsp/util.lua
Comment (mfussenegger) [inline runtime/lua/vim/lsp/util.lua:2238]: get_clients supports a `method` filter that can be used instead of this new method

File: runtime/lua/vim/lsp/diagnostic.lua
Comment (MariaSolOs) [inline runtime/lua/vim/lsp/diagnostic.lua:14]: Yay for using the new string constants :D

**Comment by MariaSolOs** [comment]
For my own learning: Does this take into account dynamic registration as well?

**Comment by MariaSolOs** [comment]
> I'm not sure yet. Do you know of a good way to test dynamic registration? What servers currently support it?

I'm honestly not sure either of how to test it either lol. It seems like Solargraph supports dynamic capabilities based on [this discussion in Lazy](https://github.com/LazyVim/LazyVim/discussions/990).

File: runtime/lua/vim/lsp/diagnostic.lua
Comment (justinmk) [inline runtime/lua/vim/lsp/diagnostic.lua:416]: Can this check be added to `util._refresh` instead? Could be an option on `opts` if there is a case(?) where it makes sense to request things not supported by clients...

File: runtime/lua/vim/lsp/inlay_hint.lua
Comment (justinmk) [inline runtime/lua/vim/lsp/inlay_hint.lua:138]: same quesiton here

File: runtime/lua/vim/lsp/diagnostic.lua
Comment (catlee) [reply runtime/lua/vim/lsp/diagnostic.lua:416]: yeah, good point. I'll move this check into `util._refresh`

File: runtime/lua/vim/lsp/diagnostic.lua
Comment (catlee) [inline runtime/lua/vim/lsp/diagnostic.lua:466]: Hmm, this doesn't always seem to happen.

I think I need to check `vim.lsp.get_clients` and filter out the client that is detaching, and then see if any remain.

File: runtime/lua/vim/lsp/diagnostic.lua
Comment (justinmk) [inline runtime/lua/vim/lsp/diagnostic.lua:471]: vim.iter may be worth using here (and in inlay_hint.lua).

```lua
local clients = vim.lsp.get_clients({ bufnr = bufnr, method = ms.textDocument_diagnostic })
local remaining = vim.iter(clients):fold(0, function(acc, c)
  return (c.id == args.data.client_id) and i or i + 1
end)

if remaining == 0 then
  disable(bufnr)
end
```

Comment (stasjok) [reply runtime/lua/vim/lsp/diagnostic.lua:471]: > vim.iter may be worth using here (and in inlay_hint.lua).

Maybe like this?

```lua
local clients = vim.lsp.get_clients({ bufnr = bufnr, method = ms.textDocument_diagnostic })
if vim.iter(clients):any(function(client) return client.id ~= args.data.client_id end) then
  disable(bufnr)
end
```

Comment (catlee) [reply runtime/lua/vim/lsp/diagnostic.lua:471]: That's much nicer, thank you!

File: runtime/lua/vim/lsp/util.lua
Comment (glepnir) [inline runtime/lua/vim/lsp/util.lua:2209]:
```suggestion
  if not (opts.force or false) and #vim.lsp.get_clients({ bufnr = bufnr, method = method }) == 0 then
```

**Comment by catlee** [comment]
When testing this out with some extra logging, I noticed that we're sending multiple requests for inlay_hints / diagnostics each time the document changes.

This happens because the `LspNotify` autocmd gets executed once per client notification. In the inlay_hints / diagnostics handlers, we end up calling `util._refresh`, which calls `vim.lsp.buf_request`, which sends a request to all connected clients that support the given method.

The result is that each client gets notified N times, where N is the number of clients connected that support the given method.

**Comment by catlee** [comment]
I can think of a few solutions here:

1. Change when `LspNotify` is triggered. Instead of triggering it per client [here](https://github.com/neovim/neovim/blob/master/runtime/lua/vim/lsp.lua#L1584), trigger it after all clients have been notified. In the payload, `data.client_id` could be changed to `data.client_ids`.
2. Modify `vim.lsp.buf_request` to specify which client to send the request to
3. Introduce something like `vim.lsp.buf_request_single` to send a request to a single client
4. Implement debouncing logic inside `LspNofify` handlers. Ensure that we only call `vim.lsp.buf_request` once per X milliseconds. I'm not really a fan of this one.

Any other ideas, or thoughts on the above?

**Comment by stasjok** [comment]
> * Modify `vim.lsp.buf_request` to specify which client to send the request to
> * Introduce something like `vim.lsp.buf_request_single` to send a request to a single client

There is a `request()` method on a client object for that. For example, I think you can do something like
```lua
local client = vim.lsp.get_client_by_id(data.client_id)
if client.supports_method(...) then
  client.request(_method_, params, _, bufnr)
end
```

**Comment by catlee** [comment]
> > * Modify `vim.lsp.buf_request` to specify which client to send the request to
> > * Introduce something like `vim.lsp.buf_request_single` to send a request to a single client
> 
> There is a `request()` method on a client object for that. For example, I think you can do something like
> 
> ```lua
> local client = vim.lsp.get_client_by_id(data.client_id)
> if client.supports_method(_method_) then
>   client.request(_method_, params, _, bufnr)
> end
> ```

Yes, that may be the best solution. I was worried about needing to re-implement some of the request tracking / cancellation logic from `vim.lsp.buf_request`, but I don't think that's needed here.

**Comment by catlee** [comment]
I believe I've fixed the issue with the LSP clients being triggered multiple times per change. I adopted @stasjok's suggestion and call `client.request` directly.

I will be unable to address any questions or feedback for the next week or so.

**Comment by stasjok** [comment]
Hey, I'm not a maintainer here, so you can ignore my comments, but here is my input:

1) Is `_refresh()` helper functions are adding much value instead of just `utils._refresh(ms.textDocument_inlayHint, { bufnr=bufnr })`?
2) In `utils._refresh()` you should be able to just:
```lua
local clients = vim.lsp.get_clients({ id = opts.client_id, bufnr = bufnr, method = method })
```

File: runtime/lua/vim/lsp/util.lua
Comment (stasjok) [inline runtime/lua/vim/lsp/util.lua:2220]:
```suggestion
  local clients = vim.lsp.get_clients({ bufnr = bufnr, method = method, id = opts.client_id })
```

Comment (catlee) [reply runtime/lua/vim/lsp/util.lua:2220]: Yeah, this is much nicer, thank you!

File: runtime/lua/vim/lsp.lua
Comment (mfussenegger) [inline runtime/lua/vim/lsp.lua:2121]:
```suggestion
---@param handler? lsp-handler See |lsp-handler|
```

File: runtime/lua/vim/lsp/util.lua
Comment (mfussenegger) [inline runtime/lua/vim/lsp/util.lua:2196]:
```suggestion
---@class lsp.util.RefreshOptions
```

File: runtime/lua/vim/lsp/util.lua
Comment (mfussenegger) [inline runtime/lua/vim/lsp/util.lua:2203]:
```suggestion
---@param opts? lsp.util.RefreshOptions Options table
```

**Review by mfussenegger — APPROVED** [review]

**Comment by catlee** [comment]
Is there anything left to do before this can be merged in?

**Pull request merged**
